onnxscript/version_converter/_version_converter.py

```diff
@@ -0,0 +1,337 @@
+    return None
+
+
+class _VersionConverter:
+    opset_imports: dict[str, int]
+    model_version: int
+
+    def __init__(self, target_version: int):
+        self.target_version = target_version
+
+    def _upgrade_version(self, node: ir.Node, opset_version: int, up_conversion: bool) -> None:
+        if up_conversion is True:
+            node.version = opset_version + 1
+        else:
+            node.version = opset_version - 1
+
+    def process_node(
+        self, node: ir.Node, opset_version: int, up_conversion: bool = True
+    ) -> Replacement | None:
+        if node.domain not in self.opset_imports:
+            return None
+        adapter = registry.lookup_adapters(
+            node.domain, node.op_type, opset_version, up_conversion
+        )
+        if adapter is None:
+            return None
+        context = orp.RewriterContext()
+        output = adapter(node, context)
+        if output is not None:
+            if isinstance(output, ir.Value):
+                output = [output]
+            return Replacement(output, context.nodes)
+        return None
+
+    def replace_node(self, node: ir.Node, replacement, root: ir.Graph | ir.Function) -> None:
+        logger.debug("Replacing node: %s::%s %s", node.domain, node.op_type, node.name)
+
+        ir_convenience.replace_nodes_and_values(
+            root, node, [node], replacement.new_nodes, node.outputs, replacement.new_outputs
+        )
+
+    def visit_attribute(self, attr: ir.Attr | ir.RefAttr) -> None:
+        if isinstance(attr, ir.Attr):
+            if attr.type == ir.AttributeType.GRAPH:
+                self.visit_graph(attr.value)  # type: ignore[arg-type]
+            elif attr.type == ir.AttributeType.GRAPHS:
+                for graph in attr.value:
+                    self.visit_graph(graph)  # type: ignore[arg-type]
+
+    def visit_node(
+        self,
+        node: ir.Node,
+        root: ir.Graph | ir.Function,
+        opset_version: int,
+        up_conversion: bool = True,
+    ) -> None:
+        replacement = self.process_node(node, opset_version, up_conversion)
+        if replacement is None:
+            # No change. Process attributes.
+            for attr in node.attributes.values():
+                self.visit_attribute(attr)
+            return None
+        else:
+            self.replace_node(node, replacement, root)
+        return None
+
+    def visit_graph(self, graph: ir.Graph) -> None:
+        for node in graph:
+            if node.version is None:
+                node.version = self.model_version
+            # Iterate each node from current node version -> target version
+            # and updating node based on the correct adapter
+            # Up-conversion [ver->ver+1] or down-conversion [ver->ver-1]
+            for opset_version in range(node.version, self.target_version):
+                up_conversion = True
+                if self.target_version < opset_version:
```

Comment:
Better to move the check outside the for loop (compare node.version and self.target_version).